Product Collection: Add Taxonomy filters to sidebar settings

imanish003 opened the pull request:
> ⚠️ This PR will be merged after PR #9600, as this branch is checked out from the branch of PR #9600.

This PR adds `Taxonomy` filters to sidebar settings.

<img width="283" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/2e3afa24-f1f6-4947-b3ca-04d214a0b448">
<img width="283" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/5233c28f-6098-42be-a927-6003c3d050c4">


<!-- Reference any related issues or PRs here -->

Fixes #9364 

### Testing

1. Create a new post
2. Add Product Collection block
3. With `Product Collection` block selected, check the sidebar for the Taxonomy filter i.e., product categories, tags etc. These will be displayed as shown in images below:
<img width="283" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/2e3afa24-f1f6-4947-b3ca-04d214a0b448">
<img width="283" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/5233c28f-6098-42be-a927-6003c3d050c4">

4. Make sure that you have products with categories and tags. If not, add categories or tags to some products.
5. Now select a product category or tags. After the selection, verify that changes reflect immediately in the Editor view.
6. Save and publish the post. After publishing, you can just navigate to the Frontend and verify that the changes you made in the Editor are accurately reflected.
7. Return to the Editor and validate the `Reset all` functionality of filters:
   - Clicking the `Reset all` button resets your selected filters.
   - The `Reset all` button can be found as shown in image below:
<img width="279" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/fe6ad8f9-e9f5-4507-822c-8d42a0496ed0">


* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Add Taxonomy filters to sidebar settings

imanish003 committed:
Add columns control to product collection block editor settings

- `InspectorControls` from './inspector-controls' is now imported in `edit.tsx` and used in the returned JSX of `Edit` function.
- A new file `columns-control.tsx` is added under 'product-collection' block's 'inspector-controls' directory which exports a `ColumnsControl` component. This component uses `RangeControl` from '@wordpress/components' to control the number of columns in the product collection display layout when the layout type is 'flex'.
- The types file (`types.ts`) for 'product-collection' block is updated. The `Attributes` interface is renamed to `ProductCollectionAttributes` and the `ProductCollectionContext` interface is removed. The `ProductCollectionAttributes` now includes 'queryContext', 'templateSlug', and 'displayLayout' properties.

imanish003 committed:
Merge branch 'trunk' into 9356-product-collection-editor-settings-columns

imanish003 committed:
Refactor: Simplify Fallback Return in ColumnsControl Component

This commit simplifies the fallback return value of the ColumnsControl component. Instead of returning an empty fragment (<> </>), it now returns null when the condition isn't met. This change improves readability and aligns with best practices for conditional rendering in React.

imanish003 committed:
Feature: Add 'Order By' Control to Product Collection Inspector

This commit adds a new 'Order By' control to the product collection inspector. The control allows users to specify the order of products in a collection by various attributes such as title and date. To support this, a new component 'OrderByControl' has been created and included in the product collection inspector. Additionally, the types for 'order' and 'orderBy' attributes have been updated and exported for reuse.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Add more options to OrderBy type

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Add orderby handling on frontend & editor

The main changes include:
1. Added a new property 'isProductCollectionBlock' in the block.json to denote if a block is a product collection block.
2. In the ProductCollection PHP class, a new initialization function has been defined to hook into the WordPress lifecycle, register the block, and update the query based on this block.
3. Added methods to manage query parameters for both frontend rendering and the Editor.
4. Expanded allowed 'collection_params' for the REST API to include custom 'orderby' values.
5. Defined a function to build the query based on block attributes, filters, and global WP_Query.
6. Created utility functions to handle complex query operations such as merging queries, handling custom sort values, and merging arrays recursively.

These improvements allow for more flexible and robust handling of product collections in both the front-end display and the WordPress editor. It also extends support for custom 'orderby' values in the REST API, which allows for more advanced sorting options in product collections.

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9361-product-collection-filters-on-sale

imanish003 committed:
Add 'on sale' filter and enhance settings management in product collection block

This commit introduces several changes to the product collection block.
- First, it adds a new 'on sale' filter that can be used to display only the products that are currently on sale.
- It also refactors the settings management in the product collection block to use the experimental ToolsPanel component from WordPress, which provides a more flexible and intuitive way to manage block settings.
- It moves the 'Columns' control into the ToolsPanel, along with the 'Order by' control.
- A new utility function `setQueryAttribute` is introduced to simplify setting nested query parameters.
- The structure of the `ProductCollectionAttributes` and `ProductCollectionQuery` types have been adjusted to accommodate the changes.
- Finally, it makes corresponding changes in the PHP part to handle the new 'on sale' query parameter.

This should enhance the flexibility and user-friendliness of the product collection block.

imanish003 committed:
Merge branch 'trunk' into 9361-product-collection-filters-on-sale

imanish003 committed:
Add stock status filter to WooCommerce product collection block

This commit introduces a stock status filter to the WooCommerce product collection block.

The changes include:
1. Added the ability to filter products based on their stock status within the 'product-collection' block. A new stock status control is created within the inspector-controls of the block.
2. A new 'get_stock_status_query' function is introduced in 'ProductCollection.php' which returns a query for products depending on their stock status.

Please note that the stock status filter will only appear in the experimental build for now.

imanish003 committed:
Refactor Stock Status control of Product Collection block

This commit refactors the Stock Status control. The changes aim to improve the code organization and make the behavior of the component more explicit.

The key modifications are:
1. Moved stock status related constants and functions from `inspector-controls/utils.tsx` to `inspector-controls/constants.ts`. This is done to ensure that all constants and similar utility functions are organized in one place.
2. Updated `product-collection/index.tsx` to import `getDefaultStockStatuses` from `inspector-controls/constants` instead of `inspector-controls/utils`.
3. Updated `stock-status-control.tsx` to determine whether the stock status has value or not by comparing with the default stock statuses using `fastDeepEqual`. If the stock status control is deselected, it resets the stock status to the default statuses.

These changes do not introduce any new functionalities, but improve the readability and maintainability of the code.

imanish003 committed:
Merge branch 'trunk' into 9362-product-collection-filters-stock-status

imanish003 committed:
Add keyword search control to Product Collection block

This commit introduces a keyword search functionality to the Product Collection block. The update is aimed to provide users with more flexibility and precision in product collection queries.

Key changes:
1. Introduced a new file `keyword-control.tsx` that creates a Keyword Control component. This component includes a TextControl field that allows inputting a search keyword. The keyword search is debounced to prevent unnecessary queries during input and updates the block's attributes accordingly.
2. Modified `inspector-controls/index.tsx` to include the KeywordControl in the ToolsPanel for the block's filters.
3. Adjusted `ProductCollection.php` to include the keyword search in the product query array.

With these changes, users can now search for products by keyword in the Product Collection block.

imanish003 committed:
Resolve conflicts

imanish003 committed:
Add product attributes filter control to ProductCollection block

- This commit introduces the ability to filter products by attributes in ProductCollection block.
- A new `woocommerceAttributes` key was added to the `block.json` file and the `ProductCollectionQuery` type. Also, a new file `attributes-control.tsx` was created, providing the UI component for the attribute filter control in the editor.
- In addition, updates were made to the `ProductCollection.php` file in the backend to support filtering products by attributes, and the tax query was updated to include attribute queries.
- Lastly, the `ProductCollectionInspectorControls` was updated to include the `AttributesControl` component, thus enabling users to filter products by attributes in the block editor."`

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

thealexandrelara committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Resolve conflicts

imanish003 committed:
remove unused import of getDefaultStockStatuses

imanish003 committed:
Resolve conflicts

imanish003 committed:
Delete a duplicate file

imanish003 committed:
Resolve conflicts

imanish003 committed:
Remove console log

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Add taxonomies control to Product collection block

The primary changes include:
1. `taxQuery` field in the `ProductCollectionAttributes` was changed from a string to an object in `assets/js/blocks/product-collection/types.ts` and `assets/js/blocks/product-collection/constants.ts`, accommodating the ability to query products by taxonomy terms.

2. `assets/js/blocks/product-collection/inspector-controls/utils.tsx` was moved to `assets/js/blocks/product-collection/utils.tsx` to make it available for broader use.

3. New component `TaxonomyControls` was created in `assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx`, which is included in `assets/js/blocks/product-collection/inspector-controls/index.tsx`. This new control allows users to filter products in the block by their taxonomy terms.

4. Updated the block's inspector controls in `assets/js/blocks/product-collection/inspector-controls/index.tsx` to use the new `TaxonomyControls` component.

Please note that the TaxonomyControls component uses experimental features of WordPress's FormTokenField. As a result, a comment has been added to disable eslint warnings regarding the use of experimental APIs.

imanish003 committed:
Merge branch 'trunk' into 9364-product-collection-filters-taxonomies

Aljullu commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -139,14 +144,17 @@ public function build_query( $query, $block, $page ) {
 
-		$is_on_sale = $block_context_query['woocommerceOnSale'] ?? false;
+		$is_on_sale       = $block_context_query['woocommerceOnSale'] ?? false;
+		$taxonomies_query = $this->get_filter_by_taxonomies_query( $query['tax_query'] );
```
I'm getting this PHP warning:

```
Warning: Undefined array key "tax_query" in /wp-content/plugins/woocommerce-blocks/src/BlockTypes/ProductCollection.php` on line 148
```

imanish003 replied:
Oh, yes. Good catch! I have fixed it in commit [e049cf8](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/e049cf891fdda1060f997b83b671bdfe142b4509). Just curious, where do you see these warnings? I use the Query Monitor plugin 🙂

Aljullu replied:
I see them directly on the screen. I think you need to set these values in `wp-config.php`:

```PHP
define( 'WP_DEBUG', true );
define( 'SCRIPT_DEBUG', true );
```

Aljullu commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -448,4 +456,86 @@ private function merge_tax_queries( ...$queries ) {
+	 *
+	 * For example:
+	 * User could provide "Product Categories" using "Filters" ToolsPanel available in Inspector Controls.
+	 * We use this function to extract it's query from $tax_query.
```
I think there is a typo here: "_We use this function to extract **its** query from $tax_query_".

imanish003 replied:
I have fixed it in commit [e049cf8](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/e049cf891fdda1060f997b83b671bdfe142b4509)

Aljullu commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -448,4 +456,86 @@ private function merge_tax_queries( ...$queries ) {
+	 *            )
+	 *    )
+	 *
+	 * For product categories, taxonomy would be "product_tag"
```
Should this say "_For product **tags**_"?

imanish003 replied:
I have fixed it in commit [e049cf8](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/e049cf891fdda1060f997b83b671bdfe142b4509)

Aljullu reviewed (commented):
This is looking great as well, good job @imanish003!

Similar to #9630, _Reset All_ is not working for me.

I also think there should be some more margin between `Separate with commas or the Enter key.` and `Product tags`, otherwise it looks too close:

![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/49413651-0154-416e-93dc-4534455c4592)

Besides that, I left some minor comments inline. :point_down:

imanish003 committed:
Address PR feedback & other improvements

1. Added `woocommerceAttributes` to `DEFAULT_FILTERS` in the `product-collection/constants.ts` file to fix `reset all` button issue.

2. Refactored `attributes-control.tsx` to make it more maintainable:
   - The constant `EDIT_ATTRIBUTES_URL` now uses `ADMIN_URL` from `@woocommerce/settings` for a more dynamic URL generation.
   - The interface `Props` has been renamed to `AttributesControlProps` for more explicit naming.
   - Removed the usage of `useState` and `useEffect` for selected attributes. Instead, `selectedAttributes` is now directly derived from `woocommerceAttributes`.
   - The CSS class `woocommerce-product-query-panel__external-link` was renamed to `wc-product-collection-panel__external-link`

3. Deleted the `product-collection/inspector-controls/constants.ts` file which was no longer necessary due to changes in product collection implementation.

These changes contribute to codebase quality, improving readability and maintainability.

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Merge branch '9363-product-collection-filters-attributes' of https://github.com/woocommerce/woocommerce-blocks into 9364-product-collection-filters-taxonomies

imanish003 committed:
Address PR review comments

This commit involves a significant refactoring of the default product query inside the 'product-collection/constants.ts' file. A new constant `DEFAULT_QUERY` has been defined and used to replace the previously hard-coded default query settings. This refactoring aids in code readability and future modifications.

Changes also include adjustments in 'product-collection/inspector-controls' to enhance UI/UX. A new SCSS file 'editor.scss' has been created for custom styles related to the editor.

Additions include:
- Adding a class name `product-collection-inspector-toolspanel__filters` to ToolsPanel for additional styling.
- The experimental property `__experimentalShowHowTo` is set to false for `FormTokenField` and `StockStatusControl`, to hide some additional information.

In 'product-collection/inspector-controls/taxonomy-controls.tsx', the classname `product-collection-inspector__taxonomy-control` is added for improved CSS targeting.

imanish003 commented:
Hey @Aljullu,

> Similar to https://github.com/woocommerce/woocommerce-blocks/pull/9630, Reset All is not working for me.
> I also think there should be some more margin between Separate with commas or the Enter key. and Product tags, otherwise it looks too close:

Fixed in e049cf8. Could you please review the changes once more? I greatly appreciate your valuable feedback. Thank you! 🙌🏻

imanish003 committed:
Add wc-block-editor prefix to className

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 committed:
Merge branch '9363-product-collection-filters-attributes' of https://github.com/woocommerce/woocommerce-blocks into 9364-product-collection-filters-taxonomies

imanish003 committed:
Resolve conflicts

imanish003 committed:
Add wc-block-editor- prefix with classNames

Aljullu commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -487,7 +490,51 @@ function ( $carry, $item ) {
+	 * @return array Query to filter products by taxonomies.
+	 */
+	private function get_filter_by_taxonomies_query( $tax_query ): array {
+		if ( ! isset( $tax_query ) || ! is_array( $tax_query ) ) {
```
Nit: could we simply call `is_array()` and skip the `isset()` check? `is_array` would already return false if `$tax_query` is not set, no? Or do you think the code is clearer this way?

imanish003 replied:
Good catch! I will make the changes. Thanks for pointing this out 🙌🏻

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx:
```diff
@@ -0,0 +1,265 @@
+	 * to match user expectations, due to `FormTokenField` behaviour that shows
+	 * suggestions which are case insensitive.
+	 *
+	 * Although WP tries to discourage users to add terms with the same name (case insensitive),
+	 * it's still possible if you manually change the name, as long as the terms have different slugs.
+	 * In this edge case we always apply the first match from the terms list.
```
TIL, thanks for the explanatory comment. :clap:

imanish003 replied:
I cannot take credit for this code as I copied it from [here](https://github.com/WordPress/gutenberg/blob/305c93ff4d19cefda9ecd200d2e21b35319b2e6e/packages/block-library/src/query/edit/inspector-controls/taxonomy-controls.js). I simply added TypeScript types and made minor improvements. 🙂

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx:
```diff
@@ -0,0 +1,265 @@
+		<div className="wc-block-editor-product-collection-inspector__taxonomy-control">
+			<FormTokenField
+				label={ taxonomy.name }
+				value={ value }
```
I'm a bit confused by this, but if I'm not wrong, `value` can be directly inferred from `existingTerms`, right? In that case, we shouldn't need to store `value` in the state. Ie, we could do something like this:

```suggestion
				value={ ( existingTerms || [] ).map( ( { id, name } ) => ( {
					id,
					value: name,
				} ) ) }
```

and then get rid of the `state` value and the `useEffect()` associated with it. (I might have missed something, though :sweat_smile:)

imanish003 replied:
The reason for using useState is that when a user adds a new term, `existingTerms` becomes empty. Therefore, if we were to use `existingTerms` as you suggested, the input would become empty for a few seconds while the `existingTerms` are being fetched. However, I will take a look at this and see if any improvements can be made. Thanks for pointing this out 🙌🏻

Aljullu replied:
Oooh, you are right, I see what you mean. I wonder if we could use [`usePrevious()`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/hooks/use-previous.ts), however. IMO it seems like a better hook for this use case. So we would use `existingTerms` by default, but if it's empty, we would use `previousExistingTerms` to make sure the input field doesn't become empty. What do you think? :thinking:

imanish003 replied:
Actually, I did consider using `usePrevious()`, but the implementation would be somewhat complex. For instance, if there is one term selected and the user select another term, the value of `existingTerms` would look like this:

<img width="218" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/5d083ff0-8626-4400-8454-fffb04d9869c">

As you can see in the console.log output above. There are multiple null and empty values before the second term is added to existing terms. This complicates determining when we should use the `usePrevious` value. Do you see the point? However, I will experiment with it today to see if I can figure out how to use `usePrevious`, which may require some code refactoring.

The first thing I will try is [useEntityRecords](https://developer.wordpress.org/news/2023/05/useentityrecords-an-easier-way-to-fetch-wordpress-data/) hook.

Aljullu replied:
Just for future reference: heads-up that `usePrevious()` accepts a [`validation` function](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/hooks/use-previous.ts#L18) so you can determine whether it updates or not. So in this case, it should be possible to make it so it doesn't update in certain circumstances.

Aljullu reviewed (commented):
Thanks for updating and rebasing this PR, @imanish003! This is looking good, but I found one issue. There is the possibility that two subcategories might have the same name:

![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/037a6f01-b84b-4b83-9499-22485cc7dc5a)

When that happens, there is a JS warning in the console: `Warning: Encountered two children with the same key, Accessories`. And it's impossible for the user to know which category is which.

One idea to fix this could be to append the `slug` when there are two or more terms with the same name, as we know slugs are unique. What do you think? Another idea would be to append the parent category name, but I guess that complicates the code a lot and there is also the possibility that parent categories might have the same name, so it doesn't really solve the problem.

imanish003 committed:
Handle duplicate taxonomy names in Taxonomy controls

the taxonomy controls have been enhanced in the following ways:
1. Modified the BASE_QUERY object to include 'slug' in the '_fields' property. This will ensure that the 'slug' of the taxonomy term is fetched along with its 'id' and 'name'.

2. Added a 'slug' property to the Term type to store the 'slug' of each term.

3. Updated the useEffect hook inside the TaxonomyItem function to generate suggestions based on search results. The suggestions now include the 'slug' of a term if the term's name is not unique. This change will help users distinguish between terms with the same name but different slugs.

imanish003 committed:
Merge branch 'trunk' into 9364-product-collection-filters-taxonomies

imanish003 committed:
Remove isset() in if condition as it's unnecessary

imanish003 committed:
Refactor TaxonomyItem component for better readability

Following changes were made:

1. The useSelect hooks which were being used to fetch existing terms and search results have been moved into their own custom hooks named 'useExistingTerms' and 'useSearchResults' respectively. This simplifies the TaxonomyItem function's body and makes the hooks' purposes clearer.

2. The comments and props destructuring for the TaxonomyItem function have been moved up to make it easier to understand the function's purpose and the props it receives.

3. This refactor does not introduce any changes in functionality. It only changes how the code is organized and presented, which will make future development easier.

imanish003 committed:
Handling for duplicate term names & other improvements

This commit enhances the `TaxonomyControls` component within `product-collection` block by adding memoization and improving term uniqueness handling.

Changes:

1. Imported `useMemo` from `@wordpress/element` for memoizing certain results.

2. `getTermIdByTermValue` function has been modified to use a `termIdToNameMap` (term ids as keys and term names as values). This provides a more efficient and direct mapping for term search.

3. Introduced `useTermIdToNameMap` function, which returns a `Map` where term ids are keys and term names are values. It handles duplicate term names by appending the term slug to the name, ensuring unique term names.

4. Updated the `useExistingTerms` and `useSearchResults` to include `taxonomy` in their dependency arrays for `useSelect` hook. This will force re-computation when `taxonomy` changes.

5. Changed `TaxonomyItem` from a function declaration to a const arrow function, consistent with the rest of the codebase.

6. Updated `onTermsChange` function in `TaxonomyItem` to accommodate the changes in `getTermIdByTermValue` and the introduction of `termIdToNameMap`.

7. Replaced `Set` with a standard array for storing new term IDs in `onTermsChange`. The `Set` was unnecessary as term IDs are unique by default.

8. Updated `TaxonomyItem`'s effects and rendering to work with `termIdToNameMap`, ensuring the displayed term names are unique.

This update will result in more efficient term search and handling, and it will solve issues related to duplicate term names.

imanish003 committed:
Restructure taxonomy controls in product collection block

This commit restructures the taxonomy controls in the product collection block for improved clarity and maintainability.
- The file `taxonomy-controls.tsx` has been deleted, and its functionality has been divided into two new files: `index.tsx` and `taxonomy-item.tsx`.

- The `index.tsx` file contains the main TaxonomyControls component, which is responsible for displaying taxonomy-related options in the block's inspector controls. It includes a custom hook `useTaxonomies` that fetches and returns taxonomies associated with product post type.

- The `taxonomy-item.tsx` file, on the other hand, contains a TaxonomyItem component that handles the rendering of individual taxonomy items. It also contains some utility functions for mapping term names and ids and fetching terms based on the search query.

This refactor aims to improve code readability and separation of concerns, thus making future changes and maintenance easier.

imanish003 committed:
Merge branch 'trunk' into 9364-product-collection-filters-taxonomies

imanish003 commented:
Hey @Aljullu, I have made a good amount of changes:
- Added handling for duplicate term names
- Refactored a lot of code to improve code readability and separation of concerns, thus making future changes and maintenance easier.

Can you please take a look again? 🙂

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:
```diff
@@ -0,0 +1,257 @@
+		if ( ! searchHasResolved ) return suggestionsRef.current;
+
+		const newSuggestions = searchResults.map( ( searchResult: Term ) =>
+			termIdToNameMap.get( searchResult.id )
```
On my end, this causes a JS error when `Uncategorized` is one of the results, as it doesn't seem to be present in `termIdToNameMap` for some reason. I don't really know why. Could we add a fallback to `searchResult.name` just in case?

```suggestion
			termIdToNameMap.get( searchResult.id ) || searchResult.name
```

imanish003 replied:
I can think of only one case when this might happen, i.e. when we are editing Product Collection block in a post and in a new tab we add a new term. In that case, new term won't be available in termIdToNameMap until user refreshes the page. 

Can you think of any other case when this might happen? 

I am adding a fallback for such scenarios. 🙂

Aljullu replied:
Hmmm, I'm not sure what was going on. Tested on a different site and I can't reproduce it there. Let's leave it as-is and I will open an issue if I can reproduce again. Sorry for the confusion. :sweat:

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:
```diff
@@ -0,0 +1,257 @@
+}
+
+// Function to get the term id based on user input in the `FormTokenField`.
+export const getTermIdByTermValue = (
```
Is this function exported for some reason?

imanish003 replied:
My apologies. I accidentally left this function exported while moving it between files. There is no reason for this function to be exported. 😅

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:
```diff
@@ -0,0 +1,257 @@
+import { useDebounce } from '@wordpress/compose';
+import { FormTokenField } from '@wordpress/components';
+
+// A constant empty array that is reused throughout the component.
+const EMPTY_ARRAY: [] = [];
+
+// Base arguments for querying terms.
+const BASE_QUERY_ARGS = {
+	order: 'asc',
+	_fields: 'id,name,slug',
+	context: 'view',
+};
+
+type Term = {
+	id: number;
+	name: string;
+	slug: string;
+};
+
+interface TaxonomyItemProps {
+	taxonomy: Taxonomy;
+	termIds: number[];
+	onChange: ( termIds: number[] ) => void;
+}
```
Nit: maybe a personal taste, but IMO listing TS types _before_ JS constants makes code easier to read as TS types are not part of the file bundle but JS constants are. What do you think?

imanish003 replied:
I don't have any specific preference regarding this, but I will go ahead and move the types before the constants as you suggested. 😀

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:
```diff
@@ -0,0 +1,257 @@
+	const onTermsChange = ( newTermValues: FormTokenField.Value[] ) => {
+		const newTermIds = [];
+		for ( const termValue of newTermValues ) {
+			const termId = getTermIdByTermValue(
+				termValue as string | Term,
+				termNameToIdMap
+			);
```
I think that if you check whether `termValue` is a string or not here:

```suggestion
			const termId =
				typeof termValue === 'string'
					? getTermIdByTermValue( termValue, termNameToIdMap )
					: termValue.id;
```

Then, `getTermIdByTermValue()` can be a one-liner:

```JS
const getTermIdByTermValue = (
	searchTerm: string,
	termNameToIdMap: Map< string, number >
): number | undefined => {
	return (
		termNameToIdMap.get( searchTerm ) ||
		termNameToIdMap.get( searchTerm.toLocaleLowerCase() )
	);
};
```

imanish003 replied:
I understand your point, but I believe it's better to leave the implementation details to our abstraction, i.e. `getTermIdByTermValue`. I have made some improvements to this function based on your feedback in [d9d7cd7](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/d9d7cd77284801aad67c74a2536fc831642acb1d).

However, please let me know if you prefer checking the type outside of `getTermIdByTermValue`. I can make that change because I don't have a strong opinion on this. 🙂

Aljullu replied:
Thanks! I usually prefer functions to accept the minimum possible types as possible, so they are easier to understand, but no need to change it.

Aljullu reviewed (commented):
Thanks for updating this PR, @imanish003! I really appreciate how you split the code in two files and now it's much easier to read. I found one issue: the _Uncategorized_ category appears for me in the UI, but I'm not able to select it. Clicking on it doesn't add it to the field. Can you reproduce?

[Enregistrament de pantalla des de 2023-06-01 13-32-55.webm](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/5fc9de40-f747-4072-916c-605940e8808c)

Another thing, which is probably not a blocker, is that I feel this component makes too many calls to the API. Ie: just enabling the "Taxonomies" filter, causes three API calls. It might be nice to investigate if they can be reduced.

[Enregistrament de pantalla des de 2023-06-01 13-47-53.webm](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/5016cd48-210f-4972-b47a-2edf242e3e86)

Besides that, I left some comments inline. Some of them are code suggestions, I did test them but might have missed some edge cases, so feel free to disregard if they don't work well.

imanish003 committed:
Fix case insensitive search support for FormTokenField

This change enhances the search functionality of the FormTokenField by introducing support for case insensitive search. This has been achieved by adding a lower-case version of the term name to the 'termNameToIdMap'.

This is an important enhancement as it will make the search process more user-friendly and resilient to different casing inputs. Users will now be able to find the desired taxonomy term regardless of their input's case.

imanish003 commented:
Hey @Aljullu 👋 

> I found one issue: the Uncategorized category appears for me in the UI, but I'm not able to select it. Clicking on it doesn't add it to the field. Can you reproduce?

While I was able to add `Uncategorized`, I encountered an issue when trying to add its lowercase version, `uncategorized`. As a result, I pushed a fix in [6a10f3f](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/6a10f3fb5359ab455266c6c273d4cbd9c804f574). Perhaps that was the same problem you faced as well? Could you please check if this fix works for you too?

imanish003 committed:
Refactor getTermIdByTermValue function and update newSuggestions mapping

This commit does a couple of important things:
1. Reorders the definition of constants in `TaxonomyItemProps` for clarity.
2. Refactors the `getTermIdByTermValue` function. Instead of checking for an exact term name match in a convoluted manner, it now directly tries to fetch the `id` from the `searchTerm` if it is an object. If the `searchTerm` is not an object, the function tries to match it against the `termNameToIdMap` in both normal and lowercase forms. This simplification makes the function more readable and concise.
3. Updates the `newSuggestions` mapping in the `TaxonomyItem` component. It now has a fallback to `searchResult.name` if a term's name is not found in `termIdToNameMap`. This change ensures that even if the term's name is not in the map for some reason, we can still display a suggestion using the original name of the term.

imanish003 committed:
Merge branch 'trunk' into 9364-product-collection-filters-taxonomies

imanish003 committed:
Optimize term fetching and initial search state in TaxonomyItem

This commit introduces a couple of improvements to the TaxonomyItem component.

1. The initial state of the 'search' state variable has been updated to 'undefined'. This change helps prevent unnecessary initial fetching of terms when the search input is empty.

2. Term fetching logic has been optimized to only enable term fetching when necessary:
   a) Fetching based on the search query is only enabled when 'search' is not 'undefined'.
   b) Fetching existing terms is only enabled when there are term IDs.

3. The block of code responsible for fetching existing terms and setting the current value has been moved upwards. This reordering of code does not change the functionality, but it groups together similar pieces of code, enhancing readability and maintainability.

These optimizations make the component more efficient by reducing unnecessary requests and computations, and they improve the code organization.

imanish003 commented:
@Aljullu 
> Another thing, which is probably not a blocker, is that I feel this component makes too many calls to the API. Ie: just enabling the "Taxonomies" filter, causes three API calls. It might be nice to investigate if they can be reduced.

I have implemented the changes in [7b56958](https://github.com/woocommerce/woocommerce-blocks/pull/9634/commits/7b56958aa1c69e15ebc849707ae79c4131b6a67e). I have optimized the data fetching process, resulting in a single call. Can you please take a look?

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/taxonomy-controls/taxonomy-item.tsx:
```diff
@@ -0,0 +1,269 @@
+	}
+
+	return (
+		termNameToIdMap.get( searchTerm as string ) ||
```
After 6a10f3fb5359ab455266c6c273d4cbd9c804f574, we don't need this line anymore, no? Because `termNameToIdMap` already includes lower-case terms so we can default to it.

imanish003 replied:
I think we need it because of [this reason](https://github.com/WordPress/gutenberg/blob/a993bfd8e108561bc2ca37a13415847d87c2019f/packages/block-library/src/query/edit/inspector-controls/taxonomy-controls.js#L36-L39). no? 🧐

Aljullu replied:
Oooh, right. Ignore my comment. :sweat:

Aljullu reviewed (approved):
LGTM, thanks for working on this, @imanish003!

I left one minor suggestion and answered all open threads, but approving. :ship:
